lib/ansiblelint/rules/RunOnceRule.py

```diff
@@ -0,0 +1,67 @@
+
+# Copyright (c) 2020 Sorin Sbarnea <sorin.sbarnea@gmail.com>
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+import sys
+
+from ansiblelint.rules import AnsibleLintRule
+
+
+class RunOnceRule(AnsibleLintRule):
+    """RunOnceRule Class."""
+
+    id = "107"
+    shortdesc = "Unpredictable run_once"
+    description = (
+        "Use of run_once does not work with free execution strategy. "
+        "Avoid it, or use noqa comments to ignore it.")
+    severity = 'MEDIUM'
+    tags = ['deprecated']
+    version_added = 'v4.4.0'
+
+    def matchtask(self, file, task):
+        return 'run_once' in task
```

Comment:
Where does this rule limit to only plays that are running with `free`?

Generally speaking I'm -1 on this rule, unless it can be explicitly targeted to the `free` strategy.  It's a very commonly used, and useful keyword.  

Blanket discouraging its use is not in alignment with Ansible. ansible-lint should not be instructing users to not use functionality of Ansible, unless it's been deprecated within Ansible itself.  If ansible-lint cannot properly apply context, and someone feels like this should definitely be included, then it should not be a rule that is enabled by default, and a rule that concerned authors can enable.

Reply:
+1 for "only if it's off by default" in its current form, since there's not really a viable alternative I can think of to `run_once` in eg, role contexts running with >1 host. Even in playbook contexts, avoiding `run_once` would generally require multiple plays, one with an explicit host, which has other knock-on effects in the face of failed hosts that `run_once` already handles more gracefully. `run_once` generally works quite well under linear strategies (eg propagating the registered result to all hosts), and we're already issuing a runtime warning under `strategy: free`. A rule might be able to detect that set of conditions in the simplest cases, but for roles, includes, etc, there's no way to know for sure until runtime. So seems like "off by default" is the safest thing here, absent a better alternative to `run_once`.

Reply:
While I think it can partially avoided when strategy is mentioned explicitly, I doubt we can have any guarantee regarding now code run in production. Imagine someone uses it inside a role, you do not have any control on how users are consuming this role. 

That us where static analyser, the linter, comes in, warning the role author that use of run_one does not always work the way he would imagine. If he accepts that, he can document this aspect and bypass the rule violation with a simple comment.

If we do not due this on all/most rules, we avoid to warn the author before is too late.

Still, I can see one genuine case where we can avoid triggering the rule when strategy is mentioned at current playbook and explicitly using liner. In that particular case we should to trigger it.

We may also need to rephrase the message and make it more descriptive, explaining user which options he has. Such a rule does not mean that code is bad, it only needs confirmation that the risk was already considered.

Reply:
Where does the code check for the current strategy?

Also, `run_once` does work with `free` strategy. See documentation [here](https://docs.ansible.com/ansible/latest/user_guide/playbooks_delegation.html#run-once) and a note from that page states

_When used together with “serial”, tasks marked as “run_once” will be run on one host in each serial batch. If it’s crucial that the task is run only once regardless of “serial” mode, use when: inventory_hostname == ansible_play_hosts_all[0] construct._

Also not sure why its tagged with: `tags = ['deprecated']` ? Is something deprecated?